### PR TITLE
Add Public Minting Functionality to EvolutionCollectionPrecompile and Solidity Interface

### PR DESCRIPTION
**Pull request by magecnion**
## **Type**
Enhancement

___

## **Description**
This PR introduces public minting functionality to the `EvolutionCollectionPrecompile` and the corresponding Solidity interface. The main changes include:
- Addition of `EnablePublicMinting`, `DisablePublicMinting`, and `IsPublicMintingEnabled` actions in the `Action` enum in `lib.rs`.
- Implementation of the corresponding functions in the `Precompile` implementation for `EvolutionCollectionPrecompile`. These functions interact with the `LaosEvolution` pallet to enable, disable, and check the status of public minting.
- Addition of `PublicMintingEnabled` and `PublicMintingDisabled` events in the `EvolutionCollection.sol` Solidity interface.
- Addition of tests for the new public minting functionality, including checks for correct log generation and gas costs.

___

## **Changes walkthrough**
<table><thead><tr><th></th><th>Relevant files&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>lib.rs&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        ownership-chain/precompile/evolution-collection/src/lib.rs<br><br>

**The file `lib.rs` has been updated to include public minting <br>functionality in the `EvolutionCollectionPrecompile` struct. <br>This includes the addition of new constants for the Solidity <br>selectors of the `EnabledPublicMinting` and <br>`DisabledPublicMinting` logs. The `Action` enum has been <br>expanded to include `EnablePublicMinting`, <br>`DisablePublicMinting`, and `IsPublicMintingEnabled` <br>actions. Corresponding functions for these actions have been <br>added to the `Precompile` implementation for <br>`EvolutionCollectionPrecompile`. These functions interact <br>with the `LaosEvolution` pallet to enable, disable, and <br>check the status of public minting.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/299/files#diff-226dbc8d8b24203a2bf2dc68997bf410f845992128754b81e4e30615ab4ac0ee"> +95/-0</a></td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>EvolutionCollection.sol&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol<br><br>

**The `EvolutionCollection.sol` Solidity interface has been <br>updated to include the `enablePublicMinting`, <br>`disablePublicMinting`, and `isPublicMintingEnabled` <br>functions. Additionally, `PublicMintingEnabled` and <br>`PublicMintingDisabled` events have been added.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/299/files#diff-69e12becfadc0223aeeb0079b691c5d6c6d968d74883768d2d85d23cd7167b98"> +19/-0</a></td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>tests.rs&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </strong></summary>
      <ul>
        ownership-chain/precompile/evolution-collection/src/tests.rs<br><br>

**The `tests.rs` file has been updated to include tests for <br>the new public minting functionality. This includes tests <br>for enabling and disabling public minting, checking if <br>public minting is enabled, and ensuring that these functions <br>generate the correct logs. Additionally, the tests also <br>check the gas costs for these new functions. Some existing <br>test function names have been updated for clarity.**
</ul>
    </details>
  </td>
  <td><a href="https://github.com/freeverseio/laos/pull/299/files#diff-1d737a09695f0303307ccdc6d8e28354d038c837d79c4327406dae465400790d"> +136/-10</a></td>

</tr>                    
</table></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol [hunk 1]
```diff
@@ -57,4 +63,17 @@ interface EvolutionCollection {
         uint256 _tokenId,
         string calldata _tokenURI
     ) external;
+
+    /// @notice Enables public minting for the collection
```
Comment (tonimateos) [inline ownership-chain/precompile/evolution-collection/contracts/EvolutionCollection.sol:67]: explain below: 
// When enabled, any address is allowed to mint on this collection
// This does not affect evolution: only the owner of the collection can continue evolving assets

**Review by tonimateos — CHANGES_REQUESTED** [review]

**Review by asiniscalchi — DISMISSED** [review]

File: ownership-chain/precompile/evolution-collection/src/lib.rs [hunk 2]
```diff
@@ -215,6 +231,85 @@ where
...
+					.record(handle)?;
+
+				// Record EVM cost
+				handle.record_cost(GasCalculator::<Runtime>::weight_to_gas(consumed_weight))?;
+
+				// Record Substrate related costs
+				// TODO: Add `ref_time` when precompiles are benchmarked
+				handle.record_external_cost(None, Some(consumed_weight.proof_size()))?;
+
+				Ok(succeed(sp_std::vec![]))
```
Comment (dastanfv) [inline ownership-chain/precompile/evolution-collection/src/lib.rs:291]: should this be with `EvmDataWriter` to be more consistent?

File: ownership-chain/precompile/evolution-collection/src/lib.rs [hunk 3]
```diff
@@ -215,6 +231,85 @@ where
...
```
Comment (dastanfv) [inline ownership-chain/precompile/evolution-collection/src/lib.rs:260]: `EvmDataWriter` instead

**Review by tonimateos — APPROVED** [review]

**Pull request merged**